### PR TITLE
Documentation for writing unit tests

### LINKED ISSUE #1513: Improve build system documentation
We have received some questions regarding the usage of the CMake-based build system:

* How are back-ends enabled?
* How are test cases enabled?

We should document this better.

### PR DESCRIPTION
**Pull request by tonydp03**
This PR adds a draft documentation for writing unit tests commenting a simple example line by line and describing a sample workflow. It can be improved by adding comments to more complex examples.

### CODE REVIEW

File: docs/source/dev/test.rst [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+Writing a unit test
+==================
+
+After implementing a new functionality in Alpaka, it is recommended to test it. Indeed, having no compile time errors
+when building alpaka with any backend doesn't mean that such functionality is well implemented and behaves as expected.
+
+Unit tests are written and integrated with `Catch2`_ and they are standalone executables located in the ``test/unit`` and
+``test/integ`` paths inside the ``build`` folder. They can be built all at once by setting the ``CMake``
+argument ``-DBUILD_TESTING=ON`` and running the command ``cmake --build .``, and then run with the ``ctest`` command as
+described in the :doc:`Installation section </basic/install>`.
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]: Consider pointing out that the commands are supposed to be run in the build folder.

File: docs/source/dev/test.rst [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+.. _Catch2: https://github.com/catchorg/Catch2
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]: Maybe just put the link behind the first mentioning of Catch2 in the preceding paragraph.

File: docs/source/dev/test.rst [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    alpaka_add_executable(
+        ${_TARGET_NAME}
+        ${_FILES_SOURCE})
+    target_link_libraries(
+        ${_TARGET_NAME}
+        PRIVATE common)
+
+    set_target_properties(${_TARGET_NAME} PROPERTIES FOLDER "test/unit")
+
+    add_test(NAME ${_TARGET_NAME} COMMAND ${_TARGET_NAME} ${_ALPAKA_TEST_OPTIONS})
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]: I am hesitant to show the content of a `CMakeLists.txt` because it does not add a lot of value and might change at some point. And users are free to implement their `CMakeLists.txt` differently.

File: docs/source/dev/test.rst [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+where the first line sets the name of the test that ``CMake`` will build for execution. Furthermore, it is necessary to add
+the folder in which the test is located to the ``CMakeLists.txt`` file under the ``test/unit`` path:
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]:
```suggestion
the folder in which the test is located to ``test/unit/CMakeLists.txt``:
```

File: docs/source/dev/test.rst [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+.. code-block:: c++
+    #include <alpaka/dev/Traits.hpp>
+    #include <alpaka/test/acc/TestAccs.hpp>
+    #include <catch2/catch.hpp>
+
+The first ``include`` statement targets specifically the header in which the functionality is included; with the second one
+only accelerators corresponding to enabled backends are selected for testing, while the third one is necessary to integrate
+Catch2.
+
+One or more test cases are then introduced with the ``TEMPLATE_LIST_TEST_CASE`` macro. It takes three arguments:
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]: There might be other tests where you would want to use a different test declaration. Maybe rephrase:
```suggestion
Test cases using an alpaka accelerator are then typically introduced with the ``TEMPLATE_LIST_TEST_CASE`` macro. It takes three arguments:
```

Comment (bernhardmgruber) [inline docs/source/dev/test.rst]: Maybe also point to the Catch2 documentation here: https://github.com/catchorg/Catch2/blob/devel/docs/test-cases-and-sections.md#type-parametrised-test-cases

File: docs/source/dev/test.rst [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+Having all the necessary objects well defined, the function can be called and the result stored in an object:
+
+.. code-block:: c++
+    auto const warpExtent = alpaka::getWarpSize(dev);
+
+Finally, the condition under which the test succeed should be defined with the ``REQUIRE`` macro:
+
+.. code-block:: c++
+    REQUIRE(warpExtent > 0);
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]: There is also the `CHECK` macro, which I usually prefer because it does not abort the test when the check fails. Maybe just point to the Catch2 documentation here on how to do checks. There is a lot more: https://github.com/catchorg/Catch2/blob/devel/docs/assertions.md

File: docs/source/dev/test.rst [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+Such condition depends on what must be actually tested (i.e. the size of the warp can never be equal or less than 0).
+Furthermore, other additional conditions can be set as well:
+
+.. code-block:: c++
+    REQUIRE(std::is_same_v<decltype(warpExtent), int>);
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]: IMO that should be a `STATIC_REQUIRE`.

File: docs/source/dev/test.rst [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+which checks that the type of the warp size is integer (cannot be float).
+
+.. hint::
+    An alternative to the ``REQUIRE`` macro is the ``STATIC_REQUIRE`` macro. The difference between them is that if the
+    test fails, in the first case the failure will be notified at run time, while in the second case it will result in
+    a compile time error.
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]: I would omit this and point to the Catch2 documentation. There is so much more, like checking for exceptions, handling floating points etc.

**Review by bernhardmgruber — COMMENTED** [review]
Thank you for adding this! Still, I think the alpaka docs are not the right place to explain too much about Catch2. They have their own (great) documentation. We should just explain how alpaka uses Catch2 and this is mostly about the `TEMPLATE_LIST_TEST_CASE` and `alpaka::test::TestAccs`, which you already did. I think you can cut a few bits and add a few pointers to the Catch2 documentation. Still, great work and thank you for doing this!

**Comment by tonydp03** [comment]
Thanks for the feedbacks @bernhardmgruber , they are extremely useful :D I'll do some changes asap; I used a very simple example as there are many more complicated ones - but I guess that any external contributor would need help to write anything similar. Also pointing to several parts of the Catch2 documentation might be a very good idea.

File: docs/source/dev/test.rst (hunk 7 above)
Comment (tonydp03) [reply docs/source/dev/test.rst]: About this, I think that it might be useful to discuss it on VC with the others as well, since all the tests that have been written have e REQUIRE rather than STATIC_REQUIRE. Maybe it's worth to change it? (STATIC_REQUIRE is actually useful imho)

File: docs/source/dev/test.rst [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+Writing a unit test
+==================
+
+After implementing a new functionality in Alpaka, it is recommended to test it. Indeed, having no compile time errors
+when building alpaka with any backend doesn't mean that such functionality is well implemented and behaves as expected.
+
+.. _Catch2: https://github.com/catchorg/Catch2
+
+Unit tests are written and integrated with `Catch2`_ and they are standalone executables located in the ``test/unit`` and
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]:
```suggestion
Unit tests are written and integrated with `Catch2 <https://github.com/catchorg/Catch2>`_ and they are standalone executables located in the ``test/unit`` and
```

File: docs/source/dev/test.rst [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+``test/integ`` paths inside the ``build`` folder. They can be built all at once by setting the ``CMake``
+argument ``-DBUILD_TESTING=ON`` and running the command ``cmake --build .`` in the ``build`` folder, and then run with the
+``ctest`` command as described in the :doc:`Installation section </basic/install>`.
+
+In alternative, a target test can be built by adding the flag ``--target <testname>`` to the ``cmake --build`` command and
+then executed as a normal binary file (``./<testname>``) in the corresponding folder mentioned above.
+
+Sample workflow
+-------
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]:
```suggestion
Sample workflow
-------------------
```

**Review by bernhardmgruber — CHANGES_REQUESTED** [review]
Looks almost good. Please also squash all your commits together into one.

File: docs/source/dev/test.rst (hunk 7 above)
Comment (bernhardmgruber) [reply docs/source/dev/test.rst]: Since you have an example of `REQUIRE` (runtime test) a few lines above, I would really like to see `STATIC_REQUIRE` here as an example for a compile time test.

Comment (fwyzard) [reply docs/source/dev/test.rst]: For someone like me that does not have prior experience with Catch2, it would be useful to have a very quick introduction to the most common checks to use (`CHECK`, `REQUIRE`, `STATIC_REQUIRE`, _etc._)

File: docs/source/dev/test.rst [hunk 11]
```diff
@@ -0,0 +1,92 @@
+.. highlight:: cpp
+
+Writing a unit test
+==================
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]:
```suggestion
Writing a unit test
===================
```

File: docs/source/dev/test.rst [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+Unit tests are written and integrated with `Catch2 <https://github.com/catchorg/Catch2>`_ and they are standalone executables located in the ``test/unit`` and
+``test/integ`` paths inside the ``build`` folder. They can be built all at once by setting the ``CMake``
+argument ``-DBUILD_TESTING=ON`` and running the command ``cmake --build .`` in the ``build`` folder, and then run with the
+``ctest`` command as described in the :doc:`Installation section </basic/install>`.
+
+In alternative, a target test can be built by adding the flag ``--target <testname>`` to the ``cmake --build`` command and
+then executed as a normal binary file (``./<testname>``) in the corresponding folder mentioned above.
+
+Sample workflow
+-------------------
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]:
```suggestion
Sample workflow
---------------
```
Ok, maybe I messed that up previously. The goal here is to make the line as long as the heading's text above it.

File: docs/source/dev/test.rst [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+For example:
+
+.. code-block:: c++
+    REQUIRE(warpExtent > 0);
+
+evaluates the expression between the round brackets and if an exception is thrown, it is caught, reported, and counted
+as a failure. Such expression depends on what must be actually tested (i.e. the size of the warp can never be equal or
+less than 0).
+Furthermore, other additional expressions can be evaluated as well:
```
Comment (bernhardmgruber) [inline docs/source/dev/test.rst]:
```suggestion
Furthermore, additional, compile-time expressions can be evaluated as well:
```
I kind of prefer stating what the macro primarily does before showing it.

**Review by bernhardmgruber — COMMENTED** [review]
Final nits and we are good!

File: docs/source/dev/test.rst (hunk 10 above)
Comment (bernhardmgruber) [reply docs/source/dev/test.rst]: Sorry about that one!

**Review by bernhardmgruber — APPROVED** [review]
